doc/pyroman.8
.TH PYROMAN 8
.SH NAME
pyroman \- a firewall configuration utility
.SH SYNOPSIS
.hy 0
.na
.TP
.B pyroman
[
.B \-hvnspP
] [
.BI \-r " RULESDIR" 
] [
.BI \-t " SECONDS" 
]
.br
[
.B \-\-help
] [
.B \-\-version
] [
.B \-\-safe
] [
.B \-\-no\-act
]
.br
[
.B \-\-print
] [
.B \-\-print-verbose
] [
.BI \-\-rules= "RULESDIR"
]
.br
[
.BI \-\-timeout= "SECONDS"
] [
.B safe
]
.SH DESCRIPTION
.B pyroman
is a firewall configuration utility.
.PP
It will compile a set of configuration files to iptables statements to
setup IP packet filtering for you.
.PP
While it is not necessary for operating and using Pyroman, you should
have understood how IP, TCP, UDP, ICMP and the other commonly used
Internet protocols work and interact. You should also have understood the
basics of iptables in order to make use of the full functionality.
.PP
.B pyroman
does not try to hide all the iptables complexity from you, but tries to
provide you with a convenient way of managing a complex networks firewall.
For this it offers a compact syntax to add new firewall rules, while still
exposing access to add arbitrary iptables rules.
.SH OPTIONS
.PD 0
.TP
.BI \-r " RULESDIR," \-\-rules= "RULES "
Load the rules from directory
.I RULESDIR 
instead of the default directory (usually
.B /etc/pyroman
)
.TP
.BI \-t " SECONDS," \-\-timeout= "SECONDS "
Wait
.I SECONDS 
seconds after applying the changes for the user to type
.B OK 
to confirm he can still access the firewall. This implies
.I \-\-safe 
but allows you to use a different timeout.
.TP
.BR \-h ", " \-\-help
Print a summary of the command line options and exit.
.TP
.BR \-V ", " \-\-version
Print the version number of
.B pyroman
and exit.
.TP
.BR \-s ", " \-\-safe ", " safe
When the firewall was committed, wait 30 seconds for the user to type
.B OK
to confirm, that he can still access the firewall (i.e. the network
connection wasn't blocked by the firewall). 
Otherwise, the firewall changes will be undone, and the firewall will be
restored to the previous state. 
Use the
.BI \-\-timeout= "SECONDS"
option to change the timeout.
.TP
.BR \-n ", " \-\-no\-act
Don't actually run iptables. This can be used to check if
.B pyroman
accepts the configuration files.
.TP
.BR \-p ", " \-\-print
Instead of running iptables, output the generated rules.
.TP
.BR \-P ", " \-\-print-verbose
Instead of running iptables, output the generated rules. Each statement
will have one comment line explaining how this rules was generated. This
will usually include the filename and line number, and is useful for
debugging.
.SH CONFIGURATION
Configuration of pyroman consists of a number of files in the directory
.IR /etc/pyroman .
These files are in python syntax, although you do not need to be a python
programmer to use these rules. There is only a small number of statements
you need to know:
.TP
.B add_host
Define a new host or network
.TP
.B add_interface
Define a new interface (group)
.TP
.B add_service
Add a new service alias (note that you can always use
e.g. www/tcp to reference the www tcp service as defined in /etc/services)
.TP
.B add_nat
Define a new NAT (Network Address Translation) rule
.TP
.B allow
Allow a service, client, server combination
.TP
.B reject
Reject access for this service, client, server combination
.TP
.B drop
Drop packets for this service, client, server combination
.TP
.B add_rule
Add a rule for this service, client, server and target combination
.TP
.B iptables
Add an arbitrary iptables statement to be executed at beginning
.TP
.B iptables_end
Add an arbitrary iptables statement to be executed at the end
.TP
Detailed parameters for these functions can be looked up by caling
.nf
cd /usr/share/pyroman
pydoc ./commands.py
.fi

.SH BUGS
None known as of pyroman-0.4 release

.SH AUTHOR
.B pyroman
was written by Erich Schubert <erich@debian.org>

.SH SEE ALSO
.BR iptables (8),
.BR iptables-restore (8)
.BR iptables-load (8)


lib/pyroman.py
#Copyright (c) 2006 Erich Schubert erich@debian.org

#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:

#The above copyright notice and this permission notice shall be included in
#all copies or substantial portions of the Software.

#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
import sys, socket, signal, re
from util import Util
from iptables import Iptables
from popen2 import popen3

class Firewall:
	"""
	Main firewall class.
	Note that the current code is NOT designed to support multiple
	firewall objects, but only uses class variables.

	hostname -- automatically filled with the computers hostname
	timeout -- timeout for confirmation of the firewall by the user. 0 will disable auto-rollback
	accept -- target chain name for the accept() user command
	drop -- target chain name for the drop() user command
	reject -- target chain name for the reject() user command

	services -- hashmap of known services
	hosts -- hashmap of known hosts
	interfaces -- hashmap of known interfaces
	chains -- hashmap of known chains
	nats -- list of NAT rules
	rules -- list of firewall rules
	"""
	hostname = socket.gethostname()
	# Timeout when the firewall setup will be rolled back when no OK is received.
	timeout = 0

	# Target names for the "accept", "drop" and "reject" commands
	accept = "accept"
	drop = "drop"
	reject = "reject"

	services = {}
	hosts = {}
	interfaces = {}
	chains = {}
	nats = []
	rules = []

	# for testing kernel version
	kernelversioncmd = "/bin/uname -r"
	_kernelversion = None

	def __init__(self):
		"""
		Dummy initialization function, will raise an exception!
		"""
		raise "Instanciation not supported!"

	class Error(Exception):
		"""
		Basic Exception class
		"""
		pass

	def verify():
		"""
		Verify the data inserted into the firewall
		"""
		for s in Firewall.services.values():
			s.verify()
		for h in Firewall.hosts.values():
			h.verify()
		for i in Firewall.interfaces.values():
			i.verify()
		for r in Firewall.rules:
			r.verify()
	verify = staticmethod(verify)

	def prepare():
		"""
		Prepare for generation run
		"""
		for s in Firewall.services.values():
			s.prepare()
		for h in Firewall.hosts.values():
			h.prepare()
		for i in Firewall.interfaces.values():
			i.prepare()
		for r in Firewall.rules:
			r.prepare()
	prepare = staticmethod(prepare)

	def iptables_version(min=None, max=None):
		"""
		Return iptables version or test for a minimum and/or maximum version

		min -- minimal iptables version required
		max -- maximum iptables version required
		"""
		return Iptables.version(min=min, max=max)
	
	iptables_version = staticmethod(iptables_version)

	def generate():
		"""
		Generate the rules from the specifications given
		"""
		Firewall.prepare()
		for r in Firewall.rules:
			r.generate()
		for n in Firewall.nats:
			n.generate()
	generate = staticmethod(generate)

	def calciptableslines():
		"""
		Calculate the lines to be passed to iptables
		"""
		# prepare firewall rules
		lines = []
		# collect tables
		tables = []
		for c in Firewall.chains.values():
			if not c.table in tables:
				tables.append(c.table)
		# process tables
		for t in tables:
			# try to provide some useful help info, in case some error occurs
			lines.append( ["*%s" % t, "table select statement for table %s" % t] )
			# first create all chains
			for c in Firewall.chains.values():
				if c.table == t:
					lines.append( [c.get_init(), c.loginfo] )
			# then write rules (which might -j to a table not yet created otherwise)
			for c in Firewall.chains.values():
				if c.table == t:
					for l in c.get_rules():
						lines.append(l)
			# commit after each table, try to make a useful error message possible
			lines.append(["COMMIT", "commit statement for table %s" % t ])
		return lines
	calciptableslines = staticmethod(calciptableslines)

	def rollback(savedlines):
		"""
		Rollback changes to the firewall, and report rollback success to the user

		savedlines -- saved firewall setting to be restored.
		"""
		# restore old iptables rules
		restored = Iptables.restore(savedlines)
		if restored:
			sys.stderr.write("*"*70+"\n")
			sys.stderr.write("  FIREWALL ROLLBACK FAILED.\n")
			sys.stderr.write("*"*70+"\n")
		else:
			sys.stderr.write("Firewall initialization failed. Rollback complete.\n")
	rollback = staticmethod(rollback)

	def print_rules(verbose):
		"""
		Print the calculated rules, as they would be passed to iptables.
		"""
		for line in Firewall.calciptableslines():
			if verbose:
				# print reasoning
				print "# %s" % line[1]
			print line[0]
	print_rules = staticmethod(print_rules)

	def execute_rules():
		"""
		Execute the generated rules, rollback on error.
		If Firewall.timeout is set, give the user some time to accept the
		new configuration, otherwise roll back automatically.
		"""
		def user_confirm_timeout_handler(signum, frame):
			"""
			This handler is called when the user does not confirm
			firewall changes withing the given time limit.
			The firewall will then be rolled back.
			"""
			raise Firewall.Error("Success not confirmed by user")

		lines = Firewall.calciptableslines()

		# Save old firewall.
		sys.stderr.write("Saving old firewall...\n")
		savedlines = Iptables.save()

		# now try to execute the new rules
		successful = False
		try:
			successful = Iptables.commit(lines)
			sys.stderr.write("New firewall commited successfully.\n")
			if Firewall.timeout > 0:
				sys.stderr.write("To accept the new configuration, type 'OK' within %d seconds!\n" % Firewall.timeout)
				# setup timeout
				signal.signal(signal.SIGALRM, user_confirm_timeout_handler)
				signal.alarm(Firewall.timeout)
				# wait for user input
				input = sys.stdin.readline()
				# reset alarm handling
				signal.alarm(0)
				signal.signal(signal.SIGALRM, signal.SIG_DFL)

				if not re.search("^(OK|YES)", input, re.I):
					raise Firewall.Error("Success not confirmed by user")
		except Iptables.Error, e:
			sys.stderr.write("*"*70+"\n")
			sys.stderr.write("An error occurred. Starting firewall restoration.\n")
			Firewall.rollback(savedlines)
			# show exception
			sys.stderr.write("%s\n" % e);
		except Firewall.Error, e:
			sys.stderr.write("*"*70+"\n")
			sys.stderr.write("An error occurred. Starting firewall restoration.\n")
			Firewall.rollback(savedlines)
			# show exception
			sys.stderr.write("%s\n" % e);
		except:
			sys.stderr.write("*"*70+"\n")
			sys.stderr.write("An error occurred. Starting firewall restoration.\n")
			Firewall.rollback(savedlines)
			sys.stderr.write("\nHere is the exception triggered during execution:\n")
			raise
	execute_rules = staticmethod(execute_rules)
	
	def kernel_version(min=None, max=None):
		"""
		Return kernel version or test for a minimum and/or maximum version

		min -- minimal kernel version required
		max -- maximum kernel version required
		"""
		if not Firewall._kernelversion:
			# query iptables version
			ir, iw, ie = popen3(Firewall.kernelversioncmd)
			iw.close()
			result = ir.readlines()
			ir.close()
			ie.close()
			Firewall._kernelversion = result[0].strip()
			# still no version number? - raise an exception
			if not Firewall._kernelversion:
				raise Error("Couldn't get kernel version!")
		if not min and not max:
			return Firewall._kernelversion
		if min:
			if Util.compare_versions(Firewall._kernelversion, min) < 0:
				return False
		if max:
			if Util.compare_versions(Firewall._kernelversion, max) > 0:
				return False
		return True
	kernel_version = staticmethod(kernel_version)
